<!DOCTYPE html>
<html>
    <head>
        <meta charset="utf-8">
        <meta http-equiv="X-UA-Compatible" content="IE=edge,chrome=1">
        <meta name="viewport" content="width=device-width">

        <title>MISO : User Manual</title>
        <meta name="description" content="An open-source LIMS for NGS sequencing centres">

        <link href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.5/css/bootstrap.min.css" rel="stylesheet">
        <link rel="stylesheet" href="/css/syntax.css">
        <link rel="stylesheet" href="/css/main.css">
    </head>
    <body>
<a href="https://github.com/TGAC/miso-lims"><img style="position: absolute; top: 0; right: 0; border: 0;" src="https://camo.githubusercontent.com/a6677b08c955af8400f44c6298f40e7d19cc5b2d/68747470733a2f2f73332e616d617a6f6e6177732e636f6d2f6769746875622f726962626f6e732f666f726b6d655f72696768745f677261795f3664366436642e706e67" alt="Fork me on GitHub" data-canonical-src="https://s3.amazonaws.com/github/ribbons/forkme_right_gray_6d6d6d.png"></a>
        <div class="container">
            <div class="row">
                <div id="header" class="col-sm-12">
                    <h2><a class="brand" href="/"><img src="/images/MISO.png"/>MISO</a>
    <small>An open-source LIMS for NGS sequencing centres</small>
</h2>

                </div>
            </div>

            <div class="row">
                
                
                    <div id="navigation" class="col-sm-2">
                        <ul class="nav nav-list">
    <!--<li><a href="/">Home</a></li>-->
    
        
        

        
    
        
        

        
            
                <li class="nav-header">User Guides</li>
            
            <li data-order="1"><a href="/usr/user-manual.html">User Manual</a></li>
        
    
        
        

        
            
                <li class="nav-header">Admin Guides</li>
            
            <li data-order=""><a href="/adm/external-webapp.html">Configuring the external webapp</a></li>
        
            
            <li data-order=""><a href="/adm/admin-manual.html">Administrator's Manual</a></li>
        
    
        
        

        
            
                <li class="nav-header">Developer Guides</li>
            
            <li data-order=""><a href="/dev/rest-testing.html">REST Testing</a></li>
        
            
            <li data-order=""><a href="/dev/maven-artifacts.html">Maven artifacts</a></li>
        
            
            <li data-order=""><a href="/dev/kit-tracking.html">Kit Tracking</a></li>
        
            
            <li data-order=""><a href="/dev/eclipse-code-formatting.html">Eclipse Code Formatting</a></li>
        
            
            <li data-order=""><a href="/dev/code-style.html">Code Style</a></li>
        
            
            <li data-order=""><a href="/dev/developers-manual.html">Developer's Manual</a></li>
        
    
        
        

        
            
                <li class="nav-header">Reference</li>
            
            <li data-order=""><a href="/ref/rest-api.html">REST API</a></li>
        
    
        
        

        
    
<!-- List additional links. It is recommended to add a divider
    e.g. <li class="divider"></li> first to break up the content. -->
</ul>

                    </div>

                    <div id="content" class="col-sm-10">
                        <div class="page-header">
    <h2>User Manual
        
    </h2>
</div>

<h2 id="introduction">Introduction</h2>

<p><a href="https://github.com/TGAC/miso-lims">MISO</a> (<u>M</u>anaging <u>I</u>nformation for <u>S</u>equencing <u>O</u>perations) is the Lab Information Management System developed at <a href="http://www.earlham.ac.uk/">Earlham</a>. MISO is heavily modelled upon the submission schemas specified by the <a href="http://www.ebi.ac.uk/ena/">SRA</a> and is therefore able to automatically generate and deploy the required XML and data files.</p>

<h3 id="history">History</h3>

<p>MISO has been developed by the <a href="http://www.earlham.ac.uk/davey-group">Davey group</a> at Earlham Institute since February 2010. The first version of MISO was released internally to Earlham staff on 19th Sept 2011. The first beta community release of MISO was 0.1.6, released May 31st, 2012. Since then, a small dedicated software team has provided the fixes and patches needed to support the institute.</p>

<p>In 2015, the <a href="http://labs.oicr.on.ca/genome-sequence-informatics/">Genome Sequence Informatics</a> team at OICR chose MISO as their new LIMS system. OICR did not want to fork MISO entirely, so together the team at Earlham and OICR worked out a development plan to better support multiple institutes with the same code base.</p>

<p>Presently, MISO has five full-time developers between the two institutions. All active work is visible on Github and undergoes thorough code reviews and testing before being integrated into the mainline branches. Releases are pushed weekly to Github.</p>

<h2 id="overview-of-features">Overview of Features</h2>

<ul>
  <li><strong>Track Sample and Library preparation</strong> - support for multiple technicians and many different workflows for sample and library preparation, customizable for your needs.</li>
  <li><strong>Monitor sequencing runs</strong> - output and provide notifications and metrics for different sequencing platforms, including Illumina, PacBio, 454, Solid and ONT</li>
  <li><strong>Multiple user role support</strong> – user-centric access control to designated areas Laboratory tracking – project description, sample receipt, library preparation, run construction, and barcoding</li>
  <li><strong>Lab instrument support</strong> – extensions available to use barcode printers, plate scanners and hand scanners directly in the MISO interface</li>
  <li><strong>Bioinformatics pipelines</strong> – monitoring and reporting of analytical processes, interacting directly with the computing cluster</li>
  <li><strong>Reporting</strong> – personalised and filtered accurate statistics from library preparation and sequencing processes and including graphical visualisations</li>
  <li><strong>Project tracking</strong> – through “traffic light” indicators, tree-structured status diagrams and statistics graphs and plots, MISO can easily show the status of each project and its elements by interactive tree and dendrogram, as well as calendar-based sequencer resource diagrams and run statistics</li>
  <li><strong>SRA Submission</strong> – automated packaging of sequencing data and metadata, deployment of these data to remote public repositories (i.e. EBI SRA)</li>
  <li><strong>Mature, open and extendable</strong> - in production use for several years, fully open and available, with strong support from major institutions, and a dedicated development team</li>
</ul>

<p>Coupled with the open-source ethos of community development, MISO can make for a cost-effective, highly modifiable and robust system that can benefit both the public knowledge repositories and sequencing centres of all sizes.</p>

<h2 id="community">Community</h2>

<ul>
  <li>MISO is <a href="https://github.com/TGAC/miso-lims">open-source and freely available</a> for usage and modification according to the GNU Public Licence V3 (GPL3)</li>
  <li>Ask questions or report bugs on the <a href="https://github.com/TGAC/miso-lims/issues">Github tracker</a></li>
  <li>Join the developer discussion (and occasional meme) at http://miso-lims.slack.com</li>
</ul>

<h1 id="getting-started">Getting Started</h1>

<h2 id="installation-and-running-miso">Installation and running MISO</h2>

<p>The simplest way to get MISO up and running quickly is to use <a href="https://www.docker.com/">Docker</a>. Images of the most recent MISO releases are available on Docker Hub in <a href="https://hub.docker.com/r/misolims/miso-lims/">misolims/miso-lims</a>. The Docker container is intended as a demonstration and not a permanent installation.</p>

<p>To use it:</p>

<ol>
  <li><a href="https://www.docker.com/products/docker">Install Docker 1.9.1+</a></li>
  <li><code>docker pull misolims/miso-lims</code></li>
  <li><code>docker run -p 8090:8080 -d -t misolims/miso-lims</code></li>
</ol>

<p>Navigate to <a href="http://localhost:8090">http://localhost:8090</a> to login to miso with the credentials <strong>admin/admin</strong>.</p>

<h2 id="logging-into-miso">Logging into MISO</h2>

<p>MISO ships with a default admin account after installation, using the username password combination of <strong>admin</strong>/<strong>admin</strong>. To add more users and groups please see the section below. For information regarding running MISO using the database or LDAP authentication, please see the Developer Manual.</p>

<h4 id="external-users">External users</h4>

<ul>
  <li>External users have limited access to MISO, via a restricted read-only interface, which shows the list of projects with which they are involved</li>
  <li>This interface comprises detailed information about which stage the project is currently at, and QC information about Samples, Libraries and Runs.</li>
</ul>

<h4 id="internal-users">Internal users</h4>

<ul>
  <li>Internal users can access most of the features of MISO, depending on the roles allocated to them via the internal user and group management system, or another user directory such as LDAP.</li>
</ul>

<h2 id="dashboard">Dashboard</h2>

<p>The dashboard gives you an at-a-glance view of what’s happening in MISO, as well as access to most functionality.</p>

<p><img src="/images/dashboard.png" alt="MISO dashboard" /></p>

<h3 id="home">Home</h3>

<p>The Home area represents the MISO dashboard, and includes widget lists of Projects, Runs, Samples, Library, etc, that the logged-in user is able to read. The input fields on the top of each widget can be used to search the relevant object list based on name, alias, description and barcode.</p>

<h3 id="my-account">My Account</h3>

<p>The My Account tab lets the user access areas of MISO via the sidebar navigation menu, or via the widgets in the main section of the screen. This tab also includes the user’s account settings and configuration. Administrators can see the full administration and configuration settings of MISO.</p>

<h3 id="my-projects">My Projects</h3>

<p>This tab lists all the projects available in MISO that are viewable by the logged-in user, with filtering functionality and provides two diagrams showing all projects and objects they contain in a round tree or dendogram style.</p>

<h3 id="reports">Reports</h3>

<p>Reporting is an important functionality of MISO, with different type of reports being able to be generated based on Projects, Runs, Samples, Libraries, and the resource calendar of sequencing machines.</p>

<h3 id="analysis">Analysis</h3>

<p>The analysis area provides access to the MISO analysis pipeline system, which allows custom pipelines to be started on a computing cluster via a web interface.</p>

<h3 id="help">Help</h3>

<p>Link to the user manual.</p>

<h1 id="working-with-miso">Working with MISO</h1>

<p>In MISO, Project elements are the heart of the system, and comprise many components. All elements are interconnected, based on models designed from the submission schemas specified by the SRA.</p>

<h2 id="project">Project</h2>

<p>A MISO project is a collection of studies, samples, and libraries. Each project has a <em>prefix</em> and an <em>ID</em>. The project ID becomes the suffix of the project’s name_,_ e.g. <strong>PRO1, PRO2</strong>. The project name is programatically constructed to be unique and as such is unchangeable by any user, including administrators. The URL address for viewing any project is **<a href="http://your.miso.host/miso/project/">http://your.miso.host/miso/project/</a><project_id>**.</project_id></p>

<h3 id="creating-a-project">Creating a project</h3>

<p><img src="/images/create-project.png" alt="Create Project page" /></p>

<h4 id="enter-project-details">Enter project details</h4>

<ul>
  <li>Click on the <em>Add Project</em> link at the top-right of the My Projects tab</li>
  <li>Enter project details:
    <ul>
      <li>Alias: 3-5 character LIMS Sample Prefix</li>
      <li>Description: canonical project name</li>
      <li>Progress: project status</li>
      <li>Reference genome: genome against which reads will be aligned (a subproject can have a different reference genome than its parent project).</li>
      <li>An ID and name will be auto-generated for the project upon saving.</li>
    </ul>
  </li>
  <li>Add an Issue Tracker key, project can be created by relevant JIRA issue information.</li>
  <li>Click <strong>Save</strong> in the top right corner</li>
</ul>

<h4 id="adding-user-permissions">Adding user permissions</h4>

<p><img src="/images/project-permissions.png" alt="Project Permissions" /></p>

<ul>
  <li>If necessary, click on the Permissions divider to expand it.</li>
  <li>Select the users who need view (read) and edit (write) permissions. Groups of users can also be selected, in which case view and/or edit permissions will be granted to each user in that group</li>
  <li>Click <strong>Save</strong> in the top right corner</li>
</ul>

<h3 id="adding-a-study">Adding a study</h3>
<p>You will need to create a single study for your project in order to make MISO happy.<br />
* On the Project page, click on the Studies section divider to expand it.<br />
* Hover over the Options menu and click Add new Study.<br />
* Enter study details: alias, description, study type. Click <strong>Save</strong> in top right corner.<br />
* Click on project alias at top left of Edit Study page.</p>

<h3 id="adding-a-project-overview">Adding a Project Overview</h3>

<p>A Project Overview represents the portion of work to be done on a particular number of samples for a particular collaborator. Once an overview has been added, this portion of the project work can now be tracked via the alerting system, and will be reflected in the progress indicator.</p>

<h2 id="study">Study</h2>

<p>A MISO study represents more fine-grained information about the sequencing Project. Each study has a <em>prefix</em> and an <em>ID</em>. The study ID becomes the suffix of the study name_,_ e.g. <strong>STU1, STU2</strong>. The study name is programatically constructed to be unique and as such is unchangeable by any user, including administrators. The URL address for viewing any study is <strong><a href="http://your.miso.host/miso/project/">http://your.miso.host/miso/study/</a>&lt;study_ID&gt;</strong>.</p>

<h4 id="creating-a-study">Creating a study</h4>

<ul>
  <li>A study is related to a specific Project and is tracked by a unique <strong>alias</strong>.</li>
  <li>Add in a Study description, outlining perhaps the origin and purpose of this Study.</li>
  <li>Add a Study alias.
    <ul>
      <li>A Study alias cannot contains reserved characters or spaces.</li>
    </ul>
  </li>
  <li>Specify the type of this Study.</li>
</ul>

<h4 id="edit-study">Edit Study</h4>

<p>Once a Study has been saved, the following operations are available when visiting the Edit Study page, via clicking on a Study’s <em>Edit</em> link in any relevant listing table, e.g. from the List Studies page, or the Studies table in a relevant Project.</p>

<h4 id="adding-experiments">Adding Experiments</h4>

<ul>
  <li>Studies can contain any number of sequencing Experiments.</li>
  <li>Select <em>Add Experiment</em> to start adding Experiments</li>
</ul>

<h2 id="experiment">Experiment</h2>

<p>A MISO Experiment represents design information about the sequencing experiment. Each Experiment has a <em>prefix</em> and an <em>ID</em>. The Experiment ID becomes the suffix of the Experiment’s name_,_ e.g. <strong>EXP1, EXP2</strong>.</p>

<p>Experiments are associated with Pools that are placed on Run lanes, which contain the actual sequencing results.</p>

<h4 id="automatic-experiment-creation">Automatic Experiment Creation</h4>

<ul>
  <li>Whilst MISO supports creating Experiments manually (see below), it is often easier to allow MISO to generate and link an Experiment to a Pool for you.</li>
  <li>From the <em>Edit Run</em> page, you are able to select Pools to be placed on Run lanes (or chambers).</li>
  <li>When assigning a Pool to a lane, the following rules apply for Experiment creation:
    <ul>
      <li>If the Pool has no Experiment already linked</li>
      <li>If there is at least one Study available on each of the Projects represented by the Pool’s dilutions</li>
    </ul>
  </li>
  <li>If these rules are valid, drop boxes will appear for each represented Project, letting you choose the Study on which to base the automatic Experiment creation.</li>
  <li>Upon clicking <em>Select</em> for each Study, one Experiment will be automatically created and linked to the Pool.</li>
</ul>

<h4 id="add-experiment-manually">Add Experiment Manually</h4>

<ul>
  <li>An Experiment is related to a specific Study and is tracked by a unique <strong>alias</strong>.
    <ul>
      <li>An Experiment alias cannot contains reserved characters or spaces.</li>
    </ul>
  </li>
  <li>Add in an Experiment title, alias, and description.</li>
  <li>Specify the platform to which this Experiment is related.</li>
  <li>Selecting the Platform will populate a list of available Pools to link to this Experiment.</li>
  <li>Double click to add a chosen Pool to this Experiment.</li>
  <li>Double clicking again on the chosen Pool area will remove it.</li>
  <li>Permissions for each Experiment are inherited from the parent Study</li>
</ul>

<h2 id="sample">Sample</h2>

<p>A MISO sample represents the physical material received upon which sample preparation, QC and library preparations are carried out to prepare that sample for eventual sequencing. Each sample has a <em>prefix</em> and an <em>ID</em>. The sample ID becomes the suffix of the sample’s name, e.g. SAM1, SAM2.</p>

<p>There are two possible categories of samples: plain and detailed. Detailed samples extend the information recorded for plain samples and also permit deriving samples from other samples (for example, extracting DNA from a tissue to produce an analyte). MISO can be configured on startup to use the type of sample best for your institution.</p>

<h3 id="creating-a-single-sample">Creating a single sample</h3>
<p>A sample is related to a specific Project and is tracked by a unique alias and identification barcode.</p>

<p><img src="/images/create-sample.png" alt="Create single sample in MISO" /></p>

<ul>
  <li>Under <em>Tracking</em>, click <em>List Samples</em>.</li>
  <li>Click <em>Add Sample</em> in the top right hand corner.</li>
  <li>Add in a sample description, outlining perhaps the origin and eventual preparation of this sample.</li>
  <li>Add a sample alias. Sample aliases must conform to a naming convention determined by your institution, e.g. <collaborator_identifier>_S<sample_number_relative_to_parent_project>_<material_identifier>, ABC_S1_H.sapiens</material_identifier></sample_number_relative_to_parent_project></collaborator_identifier></li>
  <li>Specify a receipt date if this sample has already been received. Leave blank if this sample is to be received at a later date, i.e. when the sample delivery form has been returned to your centre.</li>
  <li>Specify a NCBI-valid species name for this sample. If enabled, MISO will check this species with the NCBI and ascribe the returned taxon ID to this sample.</li>
  <li>Specify the type of this sample.</li>
  <li>Lastly, specify whether this sample has passed QC. Please see the Sample QC section on how to add Sample QCs</li>
  <li>Click <em>Save</em> at the upper right corner of the page.</li>
</ul>

<h3 id="bulk-sample-creation">Bulk sample creation</h3>
<p>Bulk sample addition is identical to single creation, but in tabular form for ease of adding multiple samples.</p>

<ul>
  <li>To start the process, click on <em>Add Row</em> at the top of the table. An empty row will be added.</li>
  <li>Add in the required fields as you would a normal single sample input.</li>
  <li>To copy this row, click the arrow icon in the <em>Copy</em> field of the table. You will see that by default, the <strong>sample_number_relative_to_parent_project</strong> in the sample alias is incremented automatically. To turn this feature off, expand the <em>Table Options</em> section above the table, and deselect the <em>Increment Sample Aliases Automatically</em> option.</li>
  <li>Alternatively, you can specify a defined number of rows to add immediately by clicking the <em>Bulk Copy</em> button. A dialog will appear asking you how many rows to add. Choose a value, e.g. “96”, and click <em>OK</em>. This will copy the last row in the table <em>x</em> times, incrementing the <strong>sample_number_relative_to_parent_project</strong>.</li>
  <li>To quickly fill down a value in a field to other Samples, fill in the value for the chosen field in an upper row, select that row, and then click the <em>Fill Down</em> arrow button in a column header. All Samples below the chosen row will have the same value applied to that field.</li>
  <li>Click the <em>Save</em> button. You will informed of any rows that did not save.</li>
</ul>

<h3 id="sample-operations">Sample operations</h3>

<p>Once a Sample has been saved, the following operations are available when visiting the Edit Sample page, via clicking on a sample’s <em>Edit</em> link in any relevant listing table, e.g. from the List Samples page, or the Samples table in a relevant Project.</p>

<h4 id="printing-sample-barcodes">Printing Sample Barcodes</h4>

<ul>
  <li>Once a Sample has been saved, a barcode will be generated for it. This will be present in the top right of the screen when editing a Sample.</li>
  <li>Hovering over the barcode will show a menu with the <em>Print</em> option.</li>
  <li>Clicking on this will show a dialog allowing the user to select which printer with which to print the barcode.</li>
</ul>

<h4 id="adding-notes">Adding Notes</h4>

<ul>
  <li>Click on <em>Add Notes</em> under the Notes header.</li>
  <li>Add any free-text notes related to the Sample.</li>
</ul>

<h4 id="adding-sample-qcs">Adding Sample QCs</h4>

<ul>
  <li>In the Edit Sample Page, Click on <em>Add Sample QC</em> under the Option menu.</li>
  <li>Add details
    <ul>
      <li>QCed By, QC Date, QC Method and Result</li>
    </ul>
  </li>
</ul>

<h4 id="adding-libraries">Adding Libraries</h4>

<ul>
  <li>In the Edit Sample Page, Click on <em>Add Library</em> under Option of Libraries Pane.</li>
  <li>Select <em>Add Library</em> to start adding Libraries to start adding samples following the documentation below</li>
</ul>

<h3 id="receiving-samples">Receiving samples</h3>

<ul>
  <li>Visit the <em>Receive Samples</em> link in the Navigation Bar to access the form to receive samples.</li>
  <li>Either type, or beep in using a barcode reader, a sample barcode.</li>
  <li>If this sample has no previously set receipt date, it will be added to the list below the input box.</li>
  <li>You can repeat this process for as many samples as you need to receive.</li>
  <li>Once finished inputting, click <em>Save.</em></li>
</ul>

<h2 id="library">Library</h2>

<p>A MISO sample represents the first step in constructing sequenceable material from an initial Sample. A Library is then diluted down to a Dilution, and put in a Pool. Each library has a <em>prefix</em> and an <em>ID</em>. The library ID becomes the suffix of the library’s name_,_ e.g. <strong>LIB1, LIB2</strong>.</p>

<h4 id="adding-single-libraries">Adding single libraries</h4>
<p>A Library is related to a specific Sample and is tracked by a unique <strong>alias</strong> and <strong>identification</strong> <strong>barcode</strong>.</p>

<p><img src="/images/add-library.png" alt="Add library button" /></p>

<ul>
  <li>Under <em>Tracking</em>, click <em>List Samples</em>. Locate the sample that you wish to derive the library from and click on the Sample name.</li>
  <li>Scroll down to the Libraries section. Hover over the <em>Options</em> menu at the top right hand corner and click <em>Add Library</em> (see above image)</li>
  <li>Add in a library description, outlining perhaps the origin and eventual preparation of this library.</li>
  <li>The Library alias inherits from the parent Sample, but with some important changes:</li>
  <li>A Library alias must conform to the naming convention determined by your institution, e.g. <collaborator_identifier>_L<parent_sample_number>-<library_number_relative_to_parent_sample>_<material_identifier>. Thus the first Library added to ABC_S1_H.sapiens would have an alias of ABC_L1-1_H.sapiens. The second would be ABC_L1-2_H.sapiens, and so on</material_identifier></library_number_relative_to_parent_sample></parent_sample_number></collaborator_identifier></li>
  <li>The creation date is automatically set to the current date.</li>
  <li>Set the Platform that this Library is being prepared for, the Library type, Library section type, Library strategy type.</li>
  <li>If this Library is to be multiplexed with an adaptor barcode, pick the barcode from the dropdown list available for the selected platform.</li>
  <li>Check the QC Passed checkbox if the library has passed and QC stages.</li>
  <li>Permissions will be inherited from the parent Sample.</li>
  <li>Click <strong>Save</strong> in the upper right corner of the page.</li>
</ul>

<h4 id="bulk-library-creation">Bulk Library creation</h4>

<ul>
  <li>Bulk Library addition is identical to single creation, but in tabular form for ease of adding multiple Libraries.</li>
  <li>You will see the list of Samples associated with the parent Project. These form the basis for Library creation.</li>
  <li>Add in the required fields as you would a normal single Library input.</li>
  <li>Select the samples you wish to use for Library creation by clicking in the <em>Select</em> column so that the row turns blue.</li>
  <li><strong>One</strong> Library will be created for each Sample selected as such in the table.</li>
  <li>To quickly fill down a value in a field to other Libraries, fill in the value for the chosen field in an upper row, select that row, and then click the <em>Fill Down</em> arrow button in a column header. All Libraries below the chosen row will have the same value applied to that field.</li>
  <li>Click the <em>Save</em> button. You will informed of any rows that did not save.</li>
</ul>

<h4 id="library-operations">Library Operations</h4>

<p>Once a Library has been saved, the following operations are available when visiting the Edit Library page, via clicking on a Library’s <em>Edit</em> link in any relevant listing table, e.g. from the List Libraries page, or the Libraries table in a relevant Project.</p>

<h5 id="adding-notes-1">Adding Notes</h5>

<ul>
  <li>Click on <em>Add Notes</em> under the Notes header.</li>
  <li>Add any free-text notes related to the Library.</li>
</ul>

<h5 id="adding-library-qcs">Adding Library QCs</h5>

<ul>
  <li>Click on <em>Add Library QC</em> under Option menu.</li>
  <li>Add details
    <ul>
      <li>QC Date, QC Method and Result.</li>
      <li>Insert Size in base pairs.</li>
      <li>Click on <em>Add</em> to save.</li>
    </ul>
  </li>
</ul>

<h5 id="add-library-dilutions">Add Library Dilutions</h5>

<ul>
  <li>Click on <em>Add Library Dilution</em> under Option of Library Dilutions Pane.</li>
  <li>Add details
    <ul>
      <li>Creation Date and Result</li>
      <li>The ID barcode for the dilution will be generated on save</li>
      <li>Click on <em>Add</em> to save.</li>
    </ul>
  </li>
</ul>

<h5 id="constructing-pools">Constructing Pools</h5>

<ul>
  <li>Once a Library Dilution has been saved, you can use it to make up a pool</li>
  <li>Click on the <em>Construct new Pool using this dilution…</em> link to go to the Pool Wizard.</li>
  <li>Please see the Pool Wizard documentation below.</li>
</ul>

<h2 id="library-dilutions-and-pools">Library Dilutions and Pools</h2>

<p>A MISO Pool contains one or more Dilutions that are to be sequenced as part of an Experiment, in a sequencer instrument Run lane (or chamber). Pools with more than one Dilution are said to be multiplexed. Each pool has a <em>prefix</em> based on the Platform Type and an <em>ID</em>. For Illumina pools, the prefix will be <strong>IPO</strong>, for 454 <strong>LPO</strong>, for SOLiD <strong>SPO</strong>. The pool ID becomes the suffix of the pool name_,_ e.g. <strong>IPO1, LPO270</strong>.</p>

<h3 id="create-library-dilutions">Create library dilutions</h3>

<p>A library dilution is a diluted portion of a library ready to be added to a pool. Usually just represents the process of adding a library to a pool, and is not stored in a tube of its own.</p>

<p><img src="/images/add-dilutions.png" alt="Add dilutions option" /></p>

<ul>
  <li>Click the <em>My Project</em> tab and select the appropriate project from the list</li>
  <li>Open the <em>Libraries</em> section and then select <em>Add Library Dilutions</em> from the <em>Options</em> menu (see image above)</li>
  <li>Click individual rows under the <em>select</em> header to select the libraries to be used to create library dilutions. Clicking the <em>Select</em> heading will select all the rows.</li>
  <li>Select the date of the dilution from the first selected row and choose it from the calendar. If all the dates are the same you can choose the down arrow beside the Dilution Date header to fill in the rest of the rows with the same date value.</li>
  <li>Fill in the concentration files one at a time</li>
  <li>Click <em>Save Dilutions</em> in the upper right hand corner.</li>
</ul>

<h3 id="creating-pools">Creating Pools</h3>
<p>A pool is one or more diluted libraries inside a tube that is loaded onto a sequencing container. A Pool is related to a specific Experiment and is tracked by a unique alias and identification barcode.</p>

<ul>
  <li>Click the <em>My Project</em> tab and select the appropriate project from the list</li>
  <li>Open the <em>Library Dilutions</em> section and select <em>Create Pools</em> from the Options menu.</li>
  <li>Add in a Pool alias and final concentration</li>
  <li>To facilitate making multiple pools at the same time, select <em>Remove selected dilutions</em> to the dilutions already used from the list. This doesn’t affect the dilution in the database, it just removes it from the list as a convenience so the same dilution wont be selected again.</li>
  <li>Select <em>Create New Pool</em>. The newly created pool shows up on the right.</li>
  <li>Click on the pool to fill in more information:
    <ul>
      <li>Check the Ready to Run checkbox if this Pool is ready to be run on a sequencer, i.e. no more dilutions need to be added.</li>
      <li>Add QCs</li>
      <li>Create orders</li>
    </ul>
  </li>
</ul>

<h2 id="sequencing-containers">Sequencing Containers</h2>

<p>Sequencing Containers represent a collection of Lanes (representing lanes, chambers, etc) that are described by a parent Run object. Containers can be populated prior to a Run being entered manually or being imported by the Notification System. This means Containers can be barcoded and entered into MISO with their Pools populated ahead-of-time so that when Runs are added manually or automatically, the Containers can either be linked via a quick lookup or automatically, respectively.</p>

<h4 id="adding-sequencing-containers-outside-a-run">Adding Sequencing Containers (outside a Run)</h4>

<ul>
  <li>Independent Container creation is accessed from the <em>Create New Sequencing Container</em> link in the navigation area</li>
  <li>Select a Platform for the Container from the <em>Platform</em> radio buttons</li>
  <li>Select an available Sequencer Reference from the <em>Sequencer</em> dropdown</li>
  <li>Select the number of Containers you want to add for the Platform (454 is the only multiple flowcell vendor. HiSeq’s are effectively “single flowcell in twos”, i.e. two run folders are created independently even though two flowcells can be run concurrently on the machine itself).</li>
  <li>Enter in an ID for this Container. You can validate that this barcode hasn’t been used before by using the <em>Lookup</em> button.</li>
  <li>A location and validation barcode can also be entered, but are not compulsory.</li>
  <li>Double click on an empty lane and either beep in or type the barcode of the pool you wish to add to this Container</li>
  <li>Click on the <em>Save</em> button to save.</li>
</ul>

<h4 id="adding-sequencing-containers-within-a-run">Adding Sequencing Containers (within a Run)</h4>

<ul>
  <li>Containers can also be created and edited from within a Run itself and the process is very similar to the above.</li>
  <li>Select the number of Containers you want to add for the Platform (454 is the only multiple flowcell vendor. HiSeq’s are effectively “single flowcell in twos”, i.e. two run folders are created independently even though two flowcells can be run concurrently on the machine itself).</li>
  <li>Enter in an ID for this Container. You can validate that this barcode hasn’t been used before by using the <em>Lookup</em> button.</li>
  <li>A location and validation barcode can also be entered, but are not compulsory.</li>
  <li>Pools are added to Container lanes from the Pool list on the right hand side of the page by double clicking on a Pool</li>
  <li><strong>NB</strong> the chosen Pools will be added <strong>sequentially</strong> to the Container, as by this point, MISO will assess that as the run is already started, it should be fully populated.</li>
  <li>If you wish to retrospectively semi-populate a Container, use the method in the previous section (“outside a Run”).</li>
</ul>

<h2 id="run">Run</h2>

<p>A MISO Run contains the sequencing results from sequencing experiments. Each Run comprises one or more Sequencing Containers, that in turn comprise Lanes (representing lanes, chambers, etc) which hold a Pool.</p>

<h4 id="automatic-run-notification">Automatic Run Notification</h4>

<p>MISO is bundled with a separate Java application that can be run on a machine with access to sequencer run folders. This Notification module allows MISO to be informed about new Runs without any involvement from a user, making it far simpler to manage Runs within MISO. More information on this system can be found in the relevant Notification System section.</p>

<h4 id="adding-a-run-manually">Adding a Run Manually</h4>

<ul>
  <li>A Run is related to Projects via the pooled Samples placed on its lanes and is tracked by a unique <strong>alias</strong>.</li>
  <li>An ID and name will be auto-generated for the Run upon saving.</li>
  <li>Select a Platform Type.</li>
  <li>Select a Sequencer Reference
    <ul>
      <li>Once a Sequencer Reference has been selected, you will be able to select and populate the Container(s). Please follow the information in the Sequencing Containers section above.</li>
    </ul>
  </li>
  <li>A Run alias, which should be the run name specified by the sequencer system, and usually relates to a run folder of the same name on the underlying filesystem of use.</li>
  <li>A Run description, giving a brief overview of what the Run represents.</li>
  <li>The Run path, which should be the run folder exposed by the sequencer itself or an intermediary storage filesystem or cluster.</li>
  <li>Check the Paired End checkbox if the Run is paired end.</li>
  <li>Select the current Run status.</li>
  <li>Click on <em>Save</em> to save the Run.</li>
</ul>

<h4 id="run-operations">Run Operations</h4>

<p>Once a Run has been saved, either manually or via the Notification System, the following operations are available when visiting the Edit Run page, via clicking on a Run’s <em>Edit</em> link in any relevant listing table, e.g. from the List Runs page, or the Runs table in a relevant Project.</p>

<ul>
  <li>Edit alias and description</li>
  <li>Choose if the Run is paired end</li>
  <li>Setting the Status of the Run manually
    <ul>
      <li>This is also helpful if the notification system fails to accurately record the status of a Run (some platforms don’t expose this information in a coherent fashion).</li>
    </ul>
  </li>
</ul>

<h5 id="editing-sequencing-containers">Editing Sequencing Containers</h5>

<ul>
  <li>See the Sequencing Containers section above.</li>
</ul>

<h5 id="adding-notes-2">Adding Notes</h5>

<ul>
  <li>Click on <em>Add Notes</em> under the Notes header.</li>
  <li>Add any free-text notes related to the Run.</li>
</ul>

<h5 id="adding-run-qcs">Adding Run QCs</h5>

<ul>
  <li>Click on <em>Add Run QC</em> under the Option menu.</li>
  <li>Add details
    <ul>
      <li>The QC User that carried out the QC.</li>
      <li>The QC date.</li>
      <li>The QC Method relates to a conceptual or practical QC steps undertaken by one of two “actioners”, i.e. SeqOps (<em>sequencing operations</em>) or SeqInfo (<em>sequence informatics</em>). The former is carried out after the physical run has completed and assessed for quality. The latter is carried out following a primary analysis stage to assess quality.</li>
      <li>The Process Selection allows the QC to be related to a given subset of lanes, e.g. if lanes 1-4 were OK and should be processed, but lanes 5-8 were not, boxes 1-4 would be clicked to reflect this.</li>
      <li>Any other free-text information should be added in the Info field.</li>
      <li>Click on <em>Add</em> to save.</li>
    </ul>
  </li>
</ul>

<h1 id="configuring-miso">Configuring MISO</h1>

<h2 id="users-and-groups">Users and Groups</h2>

<h3 id="listing-users">Listing Users</h3>

<ul>
  <li>Click the <em>List Users</em> link in the Navigation Panel. You will see a list of currently registered Users.</li>
  <li>Click the <em>Edit</em> link to start modifying a User.</li>
  <li>Click the <em>Add User</em> link to add a new User.</li>
</ul>

<h3 id="creating-users">Creating Users</h3>

<ul>
  <li>Add the required fields in the form for the user.</li>
  <li>A user can either be <strong>Internal</strong> or <strong>External</strong>
    <ul>
      <li>An Internal user will have access to the full MISO application, depending on their permissions.</li>
      <li>An External user will have access to a special user interface of MISO that only lets that user view and complete specific elements, depending on their permissions.</li>
    </ul>
  </li>
  <li>Users can easily be added to Groups by simply checking the boxes next to their names in the Groups checklist.</li>
  <li>Click <em>Save</em> to save the User.</li>
</ul>

<h3 id="listing-groups">Listing Groups</h3>

<ul>
  <li>Click the <em>List Groups</em> link in the Navigation Panel. You will see a list of currently available Groups.</li>
  <li>Click the <em>Edit</em> link to start modifying a Group.</li>
  <li>Click the <em>Add Group</em> link to add a new Group</li>
  <li>Three groups should exist in MISO already, and they are <strong>ProjectWatchers</strong>, <strong>RunWatchers</strong>, and <strong>PoolWatchers</strong>. These groups represent the ability to watch all elements of a given type for alerting updates, and as such shouldn’t be removed.</li>
</ul>

<h3 id="creating-groups">Creating Groups</h3>

<ul>
  <li>Add a Name and Description for the new Group.</li>
  <li>Users can easily be added to Groups by simply checking the boxes next to their names in the Users checklist.</li>
  <li>Click <em>Save</em> to save the Group.</li>
</ul>

<h2 id="miso-configuration">MISO Configuration</h2>

<p>Some elements of MISO can be modified through the MISO Configuration widget in the My Account page of administrators.</p>

<h3 id="barcode-printers">Barcode Printers</h3>

<ul>
  <li>Barcode printers can be added and managed from this page.</li>
  <li>Clicking the <em>Add Printer Service</em> link will add a row to the table with relevant information for the printer to add.</li>
  <li>Type in a unique <strong>Printer Service Name</strong>, select a printer context <strong>type</strong> and options will appear for that type, then choose the element that this printer will print.</li>
  <li>To print multiple element types from the same printer, simply add separate services for each of the element types, pointing to the same host.</li>
</ul>

<h2 id="alerts">Alerts</h2>

<h3 id="my-alerts">My Alerts</h3>

<p>MISO has an alerting system whereby users can watch singular elements or whole groups of elements and receive alerts about them. Depending on the way MISO is configured, you will either receive internal MISO alerts, accessed through the My Account tab, via email, or both.</p>

<h4 id="project-alerts">Project Alerts</h4>

<ul>
  <li>To receive Project level alerts, you will need to be one of the following:
    <ul>
      <li>watching a singluar Project by selecting <em>Watch</em> in the Option menu of a Project</li>
      <li>the owner of a Project</li>
      <li>in the ProjectWatchers group</li>
    </ul>
  </li>
</ul>

<h4 id="run-alerts">Run Alerts</h4>

<ul>
  <li>To receive Run level alerts, you will need to be one of the following:
    <ul>
      <li>watching a singluar Run by selecting <em>Watch</em> in the Option menu of a Project</li>
      <li>the owner of a Run</li>
      <li>in the RunWatchers group</li>
    </ul>
  </li>
</ul>

<h4 id="pool-alerts">Pool Alerts</h4>

<ul>
  <li>To receive Pool level alerts, you will need to be one of the following:
    <ul>
      <li>be the owner of a Pool</li>
      <li>in the PoolWatchers group</li>
    </ul>
  </li>
</ul>

<h3 id="recent-activity-alerts">Recent Activity Alerts</h3>

<p>If you are and Administrator, you will be able to see the list of recently generated system level alerts. This represents a traceable timeline of events for given elements, and future versions of MISO will be able to utilise these events to build up graphical representations of project flow.</p>

<h2 id="cache-administration">Cache Administration</h2>

<h4 id="flush-all-caches">Flush All Caches</h4>

<ul>
  <li>Flushing all caches can be use to clear all the locally stored POJO objects in memory. This is useful if an administrator has changed a reference value (i.e. a Sample alias) manually in the underlying database where the original value would still be held in the in-memory cache.</li>
</ul>

<h4 id="view-cache-stats">View Cache Stats</h4>

<ul>
  <li>The statistics of each of the caches can be viewed by clicking this link.</li>
</ul>

<h4 id="regenerate-barcodes">Regenerate Barcodes</h4>

<ul>
  <li>Generates all the barcodes again in the storage area on disk from the barcode strings of each relevant barcodable object.</li>
</ul>

<h4 id="reindex-alert-managers">Reindex Alert Managers</h4>

<ul>
  <li>The MISO alerting system keeps a copy of objects watched by the alerting system. This also helps MISO in being able to form an index of these objects for fast searching. This link will reindex all objects held by this system.</li>
</ul>

<h2 id="configuring-sequencer-references">Configuring Sequencer References</h2>

<p>Sequencer References represent the link between a physical sequencer and MISO, and are essential for the input of Runs into the system.</p>

<h3 id="viewing-a-sequencer-reference">Viewing a Sequencer Reference</h3>

<ul>
  <li>The All Sequencer References page shows the currently available sequencers linked to MISO.</li>
  <li>By clicking on the <em>View</em> link of a sequencer reference, all the runs performed on that machine can be seen with corresponding name, health, date and status.</li>
</ul>

<h3 id="adding-a-sequencer-reference">Adding a Sequencer Reference</h3>

<ul>
  <li>When a new sequencer reference is to be added, click on the <em>Add Sequencer Reference</em> link in the top-right of the All Sequencer References page.</li>
  <li>A new row will be added to the table, where an ID for the instrument will be generated automatically.</li>
  <li>A unique name <strong>must</strong> be supplied for the sequencer reference, and it <strong>must be</strong> the manufacturer specified name of this instrument, e.g. for an Illumina run folder 120313_SN319_0209_AD0HPAACXX, the sequencer reference name would be SN319. If the name is incorrect here, the notification system will not be able to automatically add runs to MISO.</li>
  <li>Select a platform for the instrument</li>
  <li>Set the hostname of the machine. This will either be the head node PC of a given sequencer, or more likely the hostname of a cluster or compute area where the run folders for this instrument are stored on local disk.</li>
  <li>If the hostname is available on the network, then it will show as OK in the Available column.</li>
  <li>Add the instrument to MISO by clicking on the <em>Add</em> link.</li>
</ul>

<h3 id="editing-a-sequencer-reference">Editing a Sequencer Reference</h3>

<ul>
  <li>Editing a sequencer reference allows the administrator to change the instrument information such as the name and hostname.</li>
</ul>

<h1 id="reporting">Reporting</h1>

<p>Reporting in MISO is a very flexible feature, with reports that can be generated based on Projects, Samples, Libraries and Runs. All reports can be filtered by keywords, date range and relevant types.</p>

<h3 id="generating-reports">Generating Reports</h3>

<ol>
  <li>Reports based on different elements can be generated from separate tabs for each, i.e. Project report, Sample Report, Libraries Report and Runs Report</li>
  <li>Each type of report got different filtering options
    <ol>
      <li>All type of reports can be filtered by using keyword for name, alias and description, and dates</li>
      <li>Project reports got extra filter option for Project Progress, i.e. Active, Inactive, Pending, Cancelled, Proposed, completed, Approved, Unknown.</li>
      <li>Samples reports can be filtered based on Samples Type and Sample QC</li>
      <li>Libraries reports filters are Sequencer Platform type and Library QC</li>
      <li>Runs can be filtered based on Run Status, and Sequencers Platform type</li>
    </ol>
  </li>
  <li>Press the <em>Search</em> button, will give you a list of objects based on filters,</li>
  <li><em>Reset</em> button clears all the filters and reset the list,</li>
  <li>Select the relevant objects or select <em>All</em> check box and press <em>Generate Report</em>, the report will then be produced.</li>
</ol>

<h3 id="report-output">Report output</h3>

<p>Includes</p>

<ul>
  <li>Statistics table based on the state of Projects, Sample, Library, Runs
    <ul>
      <li>with doughnut chart(s)</li>
    </ul>
  </li>
  <li>Tabular representation of all elements display information for each element separately.
    <ul>
      <li>This can be search based on keywords</li>
    </ul>
  </li>
</ul>

<h2 id="sequencer-resource-calendar">Sequencer Resource Calendar</h2>

<ul>
  <li>The Resource Calendar shows Run information for each sequencer available in MISO.</li>
  <li>Results can be filtered:
    <ul>
      <li>
        <p>Based on Instrument</p>

        <ul>
          <li>Single Instrument or all</li>
          <li>For single instrument view in multi-flowcell instruments, e.g. HiSeq2000, both cells are shown on separate rows.</li>
        </ul>
      </li>
      <li>
        <p>Based on time</p>
        <ul>
          <li>Set time periods, i.e. Current Year, Last Year, Last Week, Last Month, Last Three Months, Last Six Months</li>
          <li>In Custom, start and end dates can be selected independently</li>
        </ul>
      </li>
    </ul>
  </li>
</ul>

<h1 id="submissions">Submissions</h1>

<p>MISO has an automated submission feature, enabling you to quickly and easily submit your sequence meta data to public repositories such as the EBI. MISO will automatically generate the required XML documents based on the information already stored in the system, and submit these to the repository. MISO can also be configured to initiate upload of your actual sequence data.</p>

<h3 id="preparing-new-submissions">Preparing new Submissions</h3>

<ul>
  <li>To prepare a new Submission, go to the <em>My Account</em> tab, and choose <em>Prepare new Submission</em>.</li>
  <li>Enter a title, alias and description for your submission in the appropriate boxes.</li>
  <li>Choose the action for the submission from the options: Add, Modify, suppress, Hold, Release, Close, Protect or Validate.</li>
  <li>You can now choose which elements you want to submit:
    <ul>
      <li>From the list of Projects, clicking on any of these will reveal a drop-down list of submittable elements.</li>
      <li>Check the boxes next to the elements you wish to submit, and click <em>Save</em>.</li>
    </ul>
  </li>
</ul>

<h3 id="previewing-validating-and-submitting-metadata">Previewing, Validating and Submitting Metadata</h3>

<ul>
  <li>Once you have saved your Submission, you will see links to either Preview, Validate or Submit your submission metadata.
    <ul>
      <li><strong>Preview</strong> will show you a condensed version of the metadata in a new browser window.</li>
      <li><strong>Validate</strong> will generate the submission XMLs and send them to the EBI validation service. Any problems with the XMLs will be displayed in the browser window - errors in Red and information in Yellow. The actual metadata will be displayed below these warnings.</li>
      <li><strong>Submit Metadata</strong> will start the metadata submission process. This will also be validated, and will fail, displaying error messages if the XML is invalid.</li>
    </ul>
  </li>
</ul>

<h3 id="submitting-sequence-data">Submitting Sequence Data</h3>

<ul>
  <li>Once you have saved your Submission, you will see a link to submit your sequence data.</li>
  <li>MISO has a number of mechanisms to find data files based on the platform and submittable elements selected.</li>
  <li>These mechanisms should be suitable for most platforms outputting data to standard locations. If not, please see the Developer Manual section on writing custom FilePathGenerators.</li>
  <li>MISO also has mechanisms for uploading data to submission endpoints, e.g. the EBI SRA. MISO ships with a built-in FTP upload feature that will connect to an FTP server and upload datafiles. We are currently working on implementing other, faster upload methods such as ASPERA.</li>
</ul>

<h3 id="list-submissions">List Submissions</h3>

<ul>
  <li>Choose <em>List Submissions</em> from the admin area to see a list of all submissions that have been created so far, along with their status. You can edit any of these by clicking the <em>Edit</em> link.</li>
</ul>

<h1 id="questions-comments-feedback-patches">Questions, comments, feedback, patches</h1>

<ul>
  <li>Ask questions or report bugs on the <a href="https://github.com/TGAC/miso-lims/issues">Github tracker</a></li>
  <li>Join the developer discussion (and occasional meme) at http://miso-lims.slack.com</li>
</ul>


                    </div>
                
            </div>

            

            <div class="row">
                <div id="footer" class="col-sm-12">
                    Documentation for <a href="https://github.com/TGAC/miso-lims">MISO</a>

                </div>
            </div>
        </div>

        <script>
            function orderNav() {
                var list,
                    section,
                    header,
                    sections = [],
                    lists = {},
                    headers = {};

                var navUl = document.querySelectorAll('#navigation ul')[0],
                    navLis = document.querySelectorAll('#navigation ul li');

                if (!navUl) return;

                for (var i = 0; i < navLis.length; i++) {
                    var order, li = navLis[i];

                    if (li.classList.contains('nav-header')) {
                        section = li.textContent || li.innerText;
                        sections.push(section);
                        headers[section] = li;
                        continue;
                    }

                    if (!lists[section]) {
                        lists[section] = [];
                    }

                    order = parseFloat(li.getAttribute('data-order'))
                    lists[section].push([order, li]);
                }

                for (var i = 0; i < sections.length; i++) {
                    section = sections[i];
                    list = lists[section].sort(function(a, b) {
                        return a[0] - b[0];
                    });

                    if (header = headers[section]) {
                        navUl.appendChild(header);
                    }
                    for (var j = 0; j < list.length; j++) {
                        navUl.appendChild(list[j][1]);
                    }
                }
            }

            if (document.querySelectorAll) orderNav();
        </script>
        
    </body>
</html>
